sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -317,16 +292,52 @@ class JacksonParser(
     row
   }
 
+  private def parseTimestamp(stringValue: String): Long = {
+    // This one will lose microseconds parts.
+    // See https://issues.apache.org/jira/browse/SPARK-10681.x
+    Try(options.timestampFormat.parse(stringValue).getTime * 1000L)
+      .getOrElse {
+        // If it fails to parse, then tries the way used in 2.0 and 1.x for backwards
+        // compatibility.
+        DateTimeUtils.stringToTime(stringValue).getTime * 1000L
+      }
+  }
+
+  private def parseDate(stringValue: String): Int = {
+    Try(DateTimeUtils.millisToDays(options.dateFormat.parse(stringValue).getTime))
+      .orElse {
+        // If it fails to parse, then tries the way used in 2.0 and 1.x for backwards
+        // compatibility.
+        Try(DateTimeUtils.millisToDays(DateTimeUtils.stringToTime(stringValue).getTime))
+      }
+      .getOrElse {
+        // In Spark 1.5.0, we store the data as number of days since epoch in string.
+        // So, we just convert it to Int.
+        stringValue.toInt
+      }
+  }
+
   /**
    * Parse an object as a Map, preserving all fields.
    */
   private def convertMap(
       parser: JsonParser,
+      keyType: DataType,
       fieldConverter: ValueConverter): MapData = {
-    val keys = ArrayBuffer.empty[UTF8String]
+    val keys = ArrayBuffer.empty[Any]
     val values = ArrayBuffer.empty[Any]
     while (nextUntil(parser, JsonToken.END_OBJECT)) {
-      keys += UTF8String.fromString(parser.getCurrentName)
+
+      val keyValue = keyType match {
+        case DateType =>
+          parseDate(parser.getCurrentName)
+        case TimestampType =>
+          parseTimestamp(parser.getCurrentName)
+        case _ =>
+          UTF8String.fromString(parser.getCurrentName)
+      }
+
+      keys += keyValue
```

Comment:
How much is it difficult to implement it in more generic way and support more types as key types? I mean creation of a key converter like `fieldConverter` and call it in the same way as for values: `keys += keyConverter.apply(parser)`.

Reply:
I originally wanted to use the existing ValueConverter but obviously that did not work as that tries to get the value from the JSON (e.g. parser.getIntValue) and we need to get the name of the field (parser.getCurrentName).

I could create a new "KeyConverter".  It would look something like this.  Shall I do that?
```
  private def parseJsonFieldName[R >: Null](parser: JsonParser)
                                           (f: PartialFunction[String, R]): R = {
    f.apply(parser.getText)
  }

  private type KeyConverter = JsonParser => AnyRef
  def makeNameConverter(keyType: DataType): KeyConverter = keyType match {
    case DateType =>
      (parser: JsonParser) => parseJsonFieldName[Int](parser) {
        case str: String => parseDate(str)
      }
    case TimestampType =>
      (parser: JsonParser) => parseJsonFieldName[Long](parser) {
        case str: String => parseTimestamp(str)
      }
    case _ =>
      (parser: JsonParser) => parseJsonFieldName[UTF8String](parser) {
        case str: String => UTF8String.fromString(str)
      }
  }
```

Reply:
I don't think that support of all types as key types must block this PR. You can support only date and timestamp types in this PR, and postpone generic solution for the near future. @HyukjinKwon @viirya What do you think of that?

Reply:
Yea, that maybe can be not blocked. However, one thing I am a bit hesitant of is if we can really support other types with a min fix, and if it's worth. Otherwise I should see if it's worth of a big change. Let me check again and left a comment soon again.

Reply:
Hi @MaxGekk , @HyukjinKwon , my thoughts are that most basic types will be supported by virtue of being converted to a string and back to their correct type.  I could create Unit Tests to show which types work.  What I think will not work is nested types (e.g. a Seq of Ints) or perhaps a BinaryType, as I can imagine that being converted to string and back may not work well.  

I can add a generic way of implementing the code, as in the code snippet above, but I don't think other types will use it.  The real use case here is when we have to do some formatting when reading and writing the key value.